### PR TITLE
cargo: Switch to workspace lints

### PR DESCRIPTION
**Pull request by lexnv**
In this PR, the recommended rust best practices are followed (since rust 1.74.0):
- declares the lints at the workspace level.
- remove a couple of lints that are now declared as unstable.

- the optimization profile of the light-client is moved at the root level

### CODE REVIEW

**Review by niklasad1 — COMMENTED** [review]
Can you bump the rust-version to v1.74.0 in the workspace Cargo.toml as well? 

Otherwise, it looks good to me and nice to re-use the lints between the crates in subxt.

File: Cargo.toml [hunk 1]
```diff
@@ -36,6 +36,29 @@ repository = "https://github.com/paritytech/subxt"
...
+path_statements = "deny"
+patterns_in_fns_without_body = "deny"
+unconditional_recursion = "deny"
+unused_allocation = "deny"
+unused_comparisons = "deny"
+unused_parens = "deny"
+while_true = "deny"
+trivial_casts = "deny"
+trivial_numeric_casts = "deny"
+unused_crate_dependencies = "deny"
```
Comment (niklasad1) [inline Cargo.toml:55]: I suggest that we remove these unused crate lints and use `cargo machete` instead

Comment (lexnv) [reply Cargo.toml]: I like this a lot! Already discovered few unused crates, with the tradeoff of some false-positive subxt dependency used by our build.rs for the testing crate! Have also added a CI step for it!

**Review by niklasad1 — APPROVED** [review]
nice :muscle:

**Review by tadeohepperle — APPROVED** [review]
This is nice!

**Pull request merged**